plugins/inputs/openstack/openstack.go

```diff
@@ -0,0 +1,1059 @@
+		tags := map[string]string{
+			"description":  project.Description,
+			"domain_id":    project.DomainID,
+			"id":           project.ID,
+			"name":         project.Name,
+			"parent_id":    project.ParentID,
+			"project_tags": strings.Join(project.Tags[:], ","),
+		}
+		fields := map[string]interface{}{
+			"is_domain": project.IsDomain,
+			"enabled":   project.Enabled,
+			"projects":  len(o.projects),
+		}
+		acc.AddFields("openstack_identity", fields, tags)
+	}
+}
+
+// accumulateHypervisors accumulates statistics from hypervisors.
+func (o *OpenStack) accumulateHypervisors(acc telegraf.Accumulator) {
+	for _, hypervisor := range o.hypervisors {
+		tags := map[string]string{
+			"cpu_vendor":              hypervisor.CPUInfo.Vendor,
+			"cpu_arch":                hypervisor.CPUInfo.Arch,
+			"cpu_model":               hypervisor.CPUInfo.Model,
+			"cpu_features":            strings.Join(hypervisor.CPUInfo.Features[:], ","),
+			"status":                  hypervisor.Status,
+			"state":                   hypervisor.State,
+			"host_ip":                 hypervisor.HostIP,
+			"hypervisor_hostname":     hypervisor.HypervisorHostname,
+			"hypervisor_type":         hypervisor.HypervisorType,
+			"hypervisor_version":      strconv.Itoa(hypervisor.HypervisorVersion),
+			"id":                      hypervisor.ID,
+			"service_host":            hypervisor.Service.Host,
+			"service_id":              hypervisor.Service.ID,
+			"service_disabled_reason": hypervisor.Service.DisabledReason,
+		}
+		fields := map[string]interface{}{
+
+			"cpu_topology_sockets": hypervisor.CPUInfo.Topology.Sockets,
+			"cpu_topology_cores":   hypervisor.CPUInfo.Topology.Cores,
+			"cpu_topology_threads": hypervisor.CPUInfo.Topology.Threads,
+			"current_workload":     hypervisor.CurrentWorkload,
+			"disk_available_least": hypervisor.DiskAvailableLeast,
+			"free_disk_gb":         hypervisor.FreeDiskGB,
+			"free_ram_mb":          hypervisor.FreeRamMB,
+			"local_gb":             hypervisor.LocalGB,
+			"local_gb_used":        hypervisor.LocalGBUsed,
+			"memory_mb":            hypervisor.MemoryMB,
+			"memory_mb_used":       hypervisor.MemoryMBUsed,
+			"running_vms":          hypervisor.RunningVMs,
+			"vcpus":                hypervisor.VCPUs,
+			"vcpus_used":           hypervisor.VCPUsUsed,
+		}
+		acc.AddFields("openstack_hypervisor", fields, tags)
+	}
+}
+
+// accumulateServers accumulates statistics about servers.
+func (o *OpenStack) accumulateServers(acc telegraf.Accumulator) {
+	var compute_hosts = map[string]map[string]string{}
+	// establishing relation between server host_id and hypervisor host_name.
+	for _, project := range o.projects {
+		compute_hosts[project.ID] = map[string]string{}
+		for _, hypervisor := range o.hypervisors {
+			h := sha256.New224()
+			h.Write([]byte(string(project.ID) + string(hypervisor.HypervisorHostname)))
+			compute_hosts[project.ID][hex.EncodeToString(h.Sum(nil))] = hypervisor.HypervisorHostname
```

Comment:
The host id of a server in Openstack is a combination of project/tenant id and hypervisor id, so to correlate I am generating the host id( hex.EncodeToString(h.Sum(nil)) ), and adding the hypervisor details(host_name) in server metrics which is masked in host id of a server.

Reply:
Yeah but you are _already_ in a map that is _unique_ to the `project_id`, so you do not need to incorporate this into your hash. So the only one left in defining your hash (the only other part written to `h`) is `hypervisor.HypervisorHostname`, so why don't you use this in the map directly?
The only reason that makes sense to me is that the hash you compute is what is done to derive `server.HostID` which you are using later during lookup. But this is a fragile thing as whenever Openstack changes the way `server.HostID` is computed this plugin will break. So I think it's better to use fields that are consistent and do not rely on knowing the internals of Openstack.

Am I missing something?

Reply:
To know on which compute/hypervisor node the server is launched on, we will need an id/name of hypervisor to be present in server struct, but we have `server.HostID` which is a combination/hash of `project_id` and `hypervisor.HypervisorHostname` ( i.e. `h.Write([]byte(string(project.ID) + string(hypervisor.HypervisorHostname)))`). So, without the hash there is no correlation between server and the hypervisor upon which it is launched.

Reply:
Maybe I'm missing something, but you have a hash consisting of `h = hash(project.ID+hypervisor.HypervisorHostname)` that you use in a `m = map[project.ID][h]`. As the hash `h` contains the project-ID it is unique there, so is the map entry (as the first map level also contains the project-ID).
So if you just use `map[project.ID][hypervisor.HypervisorHostname]` you have the same "uniqueness" as the project-ID in the former example is redundant. Alternatively you could use `map[h]` and use the hash, but this would complicate the lookup I guess.

So the question is: Why do you need to use the project-ID twice?

Reply:
Let me put it this way, if H1 and H2 are hypervisors, and P1 and P2 be projects then `server.HostID` for a server launched on H1 in P1 is hash(P1+H1) and for the a server deployed on H1 in P2 is hash(P2+H1), and similarly for server on H2 in P1 is hash(P1+H2) and for server on H2 in P2 is hash(P2+H2). The server struct consists `project.ID` and `HostID`, so to know on which Hypervisor the server is deployed on, I am creating a `map[project.ID]{HostID : hypervisor.HypervisorHostname}` to obtain the `hypervisor.HypervisorHostname` using the `server.HostID`.

Reply:
Ok, but who does set `server.HostID = hash(P1+H1)`? Openstack? If so, how do find out if they later change to `server.HostID = hash(P1+H1+somesecretsource)` or `server.HostID = some_different_hash_function(P1+H1)`?

Reply:
I have changed the flow which will add host name without need of hash.